verified_boot_app_spec.md

```diff
@@ -0,0 +1,416 @@
+set does the application default to the interactive menu interface on
+subsequent invocations. This ensures a streamlined initial experience
+focused on system provisioning, while providing full transparency and
+control afterward.
+
+## 3. Operational Phases
+
+### 3.1 Environment Initialization
+
+The application can only run before `EFI_BOOT_SERVICES.ExitBootServices()` has
+been called.
+
+Upon entry, the application uses `EFI_BOOT_SERVICES.LocateHandleBuffer()` with
+`ByProtocol` for `EFI_SIMPLE_FILE_SYSTEM_PROTOCOL`, identifying all connected
+ESP-capable volumes. Each volume is queried via `OpenVolume()` to explore file
+systems for Secure Boot material and bootloaders.
+
+### 3.2 Key Discovery on Filesystems and Default Variables
+
+In this phase, the application scans all discovered ESPs for files that may
+contain usable public keys and certificates. It searches recursively for
+files with known cryptographic formats such as `.cer`, `.der`, or `.key`,
+using simple heuristics and file extension matching.
+
+Each candidate file is parsed using an internal X.509 ASN.1 parser. For
+every successfully decoded certificate or public key, the application
+extracts metadata such as:
+
+* Subject Common Name (CN) and Organization (O)
+* Issuer CN and O
+* Validity period
+* Key type and size
+* Fingerprint
+
+These keys are collected into a staging list of potential trust
+anchors.
+
+### 3.3 Bootloader Identification and Verification
+
+The application will parse the images of the bootloaders to obtain information
+about the signatures. The application will additionally validate bootloaders
+using `EFI_BOOT_SERVICES.LoadImage()`. Bootloaders that pass signature checks
+using are listed for potential trust decisions. For images with unknown but
+verifiable keys, the application initiates a trust prompt as described in the
+next section.
+
+The information about the bootloader should include:
+
+* The EFI image device path
+* Status of image verification
+* Signature list used during verification attempts
+* Associated certificates/keys
+
+### 3.4 Trust Decision and Key Enrollment
+
+For each unknown signing key discovered during verification, the application
+prompts the user with a detailed summary, including:
+
+* Key path
+* Certificate subject and issuer fields (if available)
+* Key fingerprint (SHA-256)
+* Origin (ESP, firmware default, audit log)
+
+With explicit user consent, the application enrolls the key into the `db`
+database using `SetVariable()`. Since the application operates in either
+Audit Mode or Setup Mode, authenticated write access is not required.
+
+Only the `db` (permitted image signers) may be updated during this phase.
+The `KEK` is not modified unless the user explicitly chooses to enroll
+Microsoft's default keys. No changes to `PK` are performed until the end
+of the provisioning workflow. This restriction ensures that the platform's
+trust anchors and ownership state are not altered prematurely.
```

Comment:
When is the KEK set when not using Microsoft's default keys? I was under the impression it was mandatory for Secure Boot to work at all, since it was what signed the db, and was signed by the PK.

Reply:
KEK defines authorized certificates/keys that are allowed to update the databases (all `db*` variables). KEK is intended to be used by the OS, to give OS the possibility to update db or dbx when some revocations must happen.

KEK is not required for Secure Boot to work.